lib/resty/rediscluster.lua

```diff
@@ -159,24 +181,31 @@ local function try_hosts_slots(self, serv_list)
                 local servers = { serv_list = {} }
                 for n = 1, #slots_info do
                     local sub_info = slots_info[n]
-                    --slot info item 1 and 2 are the subrange start end slots
+                    -- slot info item 1 and 2 are the subrange start end slots
                     local start_slot, end_slot = sub_info[1], sub_info[2]
 
-                    -- generate new list of servers
+                    -- generate new list of servers, including hostnames
+                    -- from Redis 7 onwards, we optionally include IPs
                     for j = 3, #sub_info do
-                      servers.serv_list[#servers.serv_list + 1 ] = { ip = sub_info[j][1], port = sub_info[j][2] }
+                        servers.serv_list[#servers.serv_list + 1 ] = { ip = sub_info[j][1], port = sub_info[j][2] }
+                        if sub_info[j][4] and sub_info[j][4][2] ~= sub_info[j][1] then
+                            servers.serv_list[#servers.serv_list + 1 ] = { ip = sub_info[j][4][2], port = sub_info[j][2] }
+                        end
```

Comment:
It is hard to review this change, because magic numbers are used and no comments are provided about the operations that are being undergone.